README.md
# Unit

Three types: Angle, HourAngle, and Time, useful in astronomy applications.

This is a Dart port of the original Go library created by Sonia Keys and hosted at
https://github.com/soniakeys/unit. 

These types are all angle-like types.  The Time type is at least angle-related.
It has conversions to and from the other types and has a function to wrap a
Time value to the fractional part of a day.

## Motivation

This package supports https://github.com/shawnlauzon/meeus, which implements a large colection of
astronomy algorithms.

## Install

See installation instructions on [pub.dev](https://pub.dev/packages/unit#-installing-tab-)

## License

MIT

CHANGELOG.md
## 1.0.3 20 May 2020

Change return type of math operations from double to num.

## 1.0.2 15 May 2020

Fix bug in Time.operator== that causes Time == comparison to do >= instead of ==.

## 1.0.1

Add longer description in package.yaml.

## 1.0.0

- Initial port of https://github.com/soniakeys/unit, including 
Angle, HourAngle, and Time and omitting RA.


test/unit_test.dart
// License: MIT
import 'dart:math' as math;

import 'package:test/test.dart';

import 'package:unit/unit.dart';

void main() {
  group('Function tests', () {
    test('ExampleFromSexa', () {
      // Typical usage:  non-negative d, m, s, and '-' to indicate negative:
      expect(fromSexa('-', 20, 30, 0), -20.5);
      // Putting the minus sign on d is not equivalent:
      expect(fromSexa(' ', -20, 30, 0), -19.5);

      // Other combinations can give the same result though:
      expect(fromSexa(' ', -20, -30, 0), -20.5);
      expect(fromSexa(' ', -21, 30, 0), -20.5);
      expect(fromSexa(' ', -22, 90, 0), -20.5);
      expect(fromSexa('-', 22, -90, 0), -20.5);
    });

    test('ExampleFromSexaSec', () {
      expect(3600, fromSexaSec(' ', 1, 0, 0));
    });

    // Test which shows that the Dart modulo (%) operator is the same as
    // the original Go PMod function.
    test('ExampleDartMod', () {
      expect(5 % 3, 2);
      expect(-5 % 3, 1);
      expect(-2.5 % 360, 357.5);

      // A more typical use:
      expect(-2.5 % 360, 357.5);
      // Output:
      //  x      y     x % y   PMod(x, y);
      //  5      3       2          2.0
      // -5      3      -2          1.0
      // -2.5  360                357.5
    });
  });

  group('Angle tests', () {
    test('ExampleAngle.fromDeg', () {
      final a = Angle.fromDeg(180);
      expect(a.rad, 3.141592653589793);
    });

    test('ExampleAngle.fromMin', () {
      final a = Angle.fromMin(83);
      expect(a.sec, closeTo(83 * 60, .000001));
    });

    test('ExampleAngle.fromSec', () {
      final a = Angle.fromSec(83);
      expect(a.sec, closeTo(83, .000001));
    });

    test('ExampleAngleFromSexa', () {
      expect(
          Angle.fromSexa('-', 0, 32, 41).min, closeTo(-32 - 41 / 60, 0.000001));
      expect(
          Angle.fromSexa(' ', 0, 32, 41).min, closeTo(32 + 41 / 60, 0.000001));
      expect(
          Angle.fromSexa('+', 0, 32, 41).min, closeTo(32 + 41 / 60, 0.000001));
      expect(Angle.fromSexa(0, 0, 32, 41).min, closeTo(32 + 41 / 60, 0.000001));
    });

    test('ExampleAngle_Cos', () {
      final a = Angle.fromDeg(60);
      expect(a.cos(), closeTo(0.500000, .000001));
    });

    test('ExampleAngle_Deg', () {
      final a = Angle(math.pi / 2);
      expect(a.deg, 90);
    });

    test('ExampleAngle_Div', () {
      final a = Angle.fromDeg(90);
      expect(a.div(2).deg, 45);
    });

    test('ExampleAngle_HourAngle', () {
      final a = Angle.fromDeg(-30);

      expect(a.hourAngle(), HourAngle.fromHour(2));
    });

    test('ExampleAngle_Min', () {
      final a = Angle.fromSexa(' ', 1, 23, 0);
      expect(a.min, closeTo(83, 0.000001));
    });

    test('ExampleAngle_Mod1', () {
      expect(Angle.fromDeg(23).mod1(), Angle.fromDeg(23));
      expect(Angle.fromDeg(383).mod1(), Angle.fromDeg(23));
      expect(Angle.fromDeg(-337).mod1(), Angle.fromDeg(23));
    });

    test('ExampleAngle_Mul', () {
      final a = Angle.fromDeg(45);
      expect(a.mul(2), Angle.fromDeg(90));
    });

    test('ExampleAngle_Rad', () {
      final a =
          Angle.fromSexa(' ', 180, 0, 0); // conversion to radians happens here
      expect(a.rad, 3.141592653589793); // no cost to access radians here
    });

    test('ExampleAngle_Sec', () {
      final a = Angle.fromSexa(' ', 0, 1, 23);
      expect(a.sec, closeTo(83.00000, 0.000001));
    });

    test('ExampleAngle_Sin', () {
      final a = Angle.fromDeg(60);
      expect(a.sin(), closeTo(0.866025, 0.000001));
    });

    test('ExampleAngle_Tan', () {
      final a = Angle.fromDeg(60);
      expect(a.tan(), closeTo(1.732051, 0.000001));
    });

    test('ExampleAngle_Time', () {
      final a = Angle.fromDeg(-30);
      expect(a, Angle.fromSexa(0, -30, 0, 0));
      expect(a.time(), Time.fromHour(-2));
    });
  });

  group('Hour Angle tests', () {
    test('ExampleHourAngle.fromHour', () {
      final h = HourAngle.fromHour(12);
      expect(h.rad, 3.141592653589793);
      expect(h.hour, 12);
      expect(HourAngle.fromHour(-2).hour, -2);
    });

    test('ExampleHourAngle.fromMin', () {
      expect(HourAngle.fromMin(-83), HourAngle.fromSexa('-', 1, 23, 0));
    });

    test('ExampleHourAngle.fromSec', () {
      expect(HourAngle.fromSec(-83), HourAngle.fromSexa('-', 0, 1, 23));
    });

    test('ExampleHourFromSexa', () {
      final h = HourAngle.fromSexa(' ', 0, 32, 41);
      expect(h.sec, closeTo(41 + 32 * 60, 0.00001));
    });

    test('ExampleHourAngle_Angle', () {
      final h = HourAngle.fromHour(-2);
      expect(h.angle(), Angle.fromDeg(-30));
    });

    test('ExampleHourAngle_Cos', () {
      final h = HourAngle.fromHour(4);
      expect(h.cos(), closeTo(0.5, 0.000001));
    });

    test('ExampleHourAngle_Hour', () {
      final h = HourAngle.fromSexa(' ', 1, 30, 0);
      expect(h.hour, 1.5);
    });

    test('ExampleHourAngle_Div', () {
      final h = HourAngle.fromHour(6);
      expect(h.div(2), HourAngle.fromHour(3));
    });

    test('ExampleHourAngle_Min', () {
      final h = HourAngle.fromSexa(' ', 1, 23, 0);
      expect(h.min, closeTo(83, 0.000001));
    });

    test('ExampleHourAngle_Mul', () {
      final h = HourAngle.fromHour(3);
      expect(h.mul(2), HourAngle.fromHour(6));
    });

    test('ExampleHourAngle_Rad', () {
      final a = HourAngle.fromSexa(
          ' ', 12, 0, 0); // conversion to radians happens here
      expect(a.rad, 3.141592653589793); // no cost to access radians here
    });

    test('ExampleHourAngle_Sec', () {
      final a = HourAngle.fromSexa(' ', 0, 1, 23);
      expect(a.sec, closeTo(83, 0.000001));
    });

    test('ExampleHourAngle_Sin', () {
      final h = HourAngle.fromHour(4);
      expect(h.sin(), closeTo(0.866025, 0.000001));
    });

    test('ExampleHourAngle_Tan', () {
      final h = HourAngle.fromHour(4);
      expect(h.tan(), closeTo(1.732051, 0.000001));
    });

    test('ExampleHourAngle_Time', () {
      final h = HourAngle.fromHour(-2);
      expect(h.hour, -2);
      expect(h.time(), Time.fromHour(-2));
    });
  });

  group('Time tests', () {
    test('ExampleTime', () {
      final t = Time(1.23);
      expect(t.toString(), '1.23');
    });

    test('ExampleFromSexa', () {
      final t = Time.fromSexa('-', 12, 34, 45.6);
      expect(t.toIso8601String(), '-12:34:45.600000');
    });

    test('ExampleTime.fromDay', () {
      final t = Time.fromDay(1);
      expect(t.sec, 86400);
      expect(t, Time.fromHour(24));
    });

    test('ExampleTime.fromHour', () {
      final t = Time.fromHour(-1);
      expect(t.sec, -3600);
      expect(t.hour, -1);
    });

    test('ExampleTime.fromMin', () {
      expect(Time.fromMin(.5), Time(30));
    });

    test('ExampleTime.fromRad', () {
      final t = Time.fromRad(math.pi);
      expect(t.hour, closeTo(12, 0.000001));
    });

    test('ExampleTime_Angle', () {
      final t = Time.fromHour(-2);
      expect(t.hour, -2);
      expect(t.angle(), Angle.fromDeg(30));
    });

    test('ExampleTime_Day', () {
      final t = Time.fromSexa(0, 48, 36, 0);
      expect(t.day, 2.025);
    });

    test('ExampleTime_Div', () {
      final t = Time.fromHour(6);
      expect(t.div(2), Time.fromHour(3));
    });

    test('ExampleTime_Hour', () {
      final t = Time.fromSexa(0, 2, 15, 0);
      expect(t.hour, 2.25);
    });

    test('ExampleTime_HourAngle', () {
      final t = Time.fromHour(-1.5);
      expect(t.hourAngle(), HourAngle.fromSexa('-', 1, 30, 0));
    });

    test('ExampleTime_Min', () {
      final t = Time.fromSexa(0, 0, 1, 30);
      expect(t.min, 1.5);
    });

    test('ExampleTime_Mod1', () {
      final t = Time.fromHour(25);
      expect(t.sec, 90000);
      expect(t.mod1(), Time(3600));
      expect(t.mod1(), Time.fromHour(1));
    });

    test('ExampleTime_Mul', () {
      final t = Time.fromHour(3);
      expect(t.mul(2), Time.fromHour(6));
    });

    test('ExampleTime_Rad', () {
      final t = Time.fromSexa(0, 12, 0, 0);
      expect(t.rad, 3.141592653589793);
    });

    test('ExampleTime_Sec', () {
      final t = Time.fromSexa(0, 0, 1, 30);
      expect(t.sec, 90);
    });
  });
}


lib/unit.dart
/// Unit defines some types used in astronomy, and heavily used by the packages
/// in github.com/shawnlauzon/meeus.
///
/// This is a Dart port of the excellent set of libraries creatd by Sonia Keys,
/// the most significant of which is github.com/soniakeys/meeus. This
/// library was ported in order to support the port of that library. The
/// original unit library is hosted at github.com/soniakeys/unit.
///
/// Most of the original Go code has been changed as little as possible, only
/// to make it match Dart semantics. Functions which are capitalized in Go are
/// made into camel case, getter functions converted to getters, and const
/// parameters are used throughout.
///
/// Three types are currently defined: Angle, HourAngle, and Time.  All are
/// commonly formatted in sexagesimal notation and the external Go package
/// github.com/soniakeys/sexagesimal has formatting
/// routines. However, this package has not been ported to Dart because it
/// is not required in github.com/shawnlauzon/meeus other than for
/// test cases. It also contains mostly Go-specific libraries and so a port to
/// Dart would be non-trivial.
///
/// The Go version of this library includes an additional type: RA. Because
/// the Dart meeus library has not yet implemented functions which require RA,
/// that type has not yet been ported.
///
/// The original types were defined as typedefs, and since these do not exist
/// in Dart, are now defined as classes with the original typedef (in all cases
/// this is num) inside the class. For Angle and HourAngle, the value is always
/// in radians.  For Time, it is seconds.
///
/// The choice of methods defined is somewhat arbitrary.  Methods were defined
/// as they were found convenient for the Meeus library, and then filled out
/// somewhat for consistency.  The convenience is often syntactic; as the
/// underlying type is num conversions to and from num are often free
/// and otherwise typically only incur a multiplication.
library unit;

export 'src/unit_base.dart';


lib/src/unit_base.dart
// License: MIT

import 'dart:math' as math;

/// Converts from parsed sexagesimal angle components to a single
/// num value.
///
/// The result is in the units of [d], the first or "largest" sexagesimal
/// component.
///
/// Typically you pass non-negative values for [d], [m], and [s], and to indicate
/// a negative value, pass '-' for [neg].  Any other value, such as ' ', '+',
/// or simply 0, leaves the result non-negative.
///
/// There are no limits on [d], [m], or [s] however.  Negative values or values
/// > 60 for m and s are allowed for example.  The segment values are
/// combined and then if neg is '-' that sum is negated.
///
/// This function would commonly be called something like DMSToDegrees, but
/// the interpretation of d as degrees is arbitrary.  The function works
/// as well on hours minutes and seconds.  Regardless of the units of [d],
/// [m] is a sexagesimal part of [d] and [s] is a sexagesimal part of [m].
num fromSexa(dynamic neg, int d, int m, num s) {
  return fromSexaSec(neg, d, m, s) / 3600;
}

/// Converts from parsed sexagesimal angle components to a single
/// num value.
///
/// The result is in the units of [s], the last or "smallest" sexagesimal
/// component.
///
/// Otherwise fromSexaSec works as [fromSexa()].
num fromSexaSec(dynamic neg, int d, int m, num s) {
  s = ((d * 60 + m) * 60) + s;
  return neg == '-' ? -s : s;
}

/// A general purpose angle.
///
/// The value is stored as radians.
///
/// There is no "AngleFromRad" constructor.  If you have a value [rad] in
/// radians, construct a corresponding Angle simply with the constructor
/// [Angle(rad)].
class Angle {
  /// The angle in radians.
  ///
  /// This is the underlying representation and involves no scaling.
  final num rad;

  /// ---------- Angle constructors: ----------

  /// Constructs an Angle value from radians.
  const Angle(this.rad);

  /// Constructs an Angle value from a value representing angular
  /// degrees where there are 360 degrees to a circle or revolution.
  ///
  /// This provides "Deg2Rad" functionality but hopefully in a more clear way.
  const Angle.fromDeg(num d) : this(d / 180 * math.pi);
  // 180 deg or pi radians in a half-circle.

  /// Constructs an Angle value from a value representing angular
  /// minutes where there are 60 minutes to a degree, and 360 degrees to a circle.
  const Angle.fromMin(num m) : this(m / 60 / 180 * math.pi);
  // 60 min in a degree, 180 deg or pi radians in a half-circle.

  /// Constructs an Angle value from a value representing angular
  /// seconds where there are 60 seconds to a minute, 60 minutes to a degree,
  /// and 360 degrees to a circle.
  const Angle.fromSec(num s) : this(s / 3600 / 180 * math.pi);
  // 3600 sec in a degree, 180 deg or pi radians in a half-circle.

  /// Constructs a new Angle value from sign, degree, minute, and second
  /// components.
  ///
  /// For [neg], pass '-' to negate the result.  Any other argument
  /// value, such as ' ', '+', or simply 0, leaves the result non-negated.
  //
  // This was the NewAngle function in the original Go code.
  Angle.fromSexa(dynamic neg, int d, int m, num s)
      : this.fromSec(fromSexaSec(neg, d, m, s));

  // ---------- Angle "getters" or conversions: ----------

  /// The angle in degrees.
  ///
  /// This provides a "Rad2Deg" functionality but hopefully in a more clear way.
  num get deg => rad * 180 / math.pi;

  /// The angle in minutes.
  num get min => rad * 60 * 180 / math.pi;

  /// The angle in seconds.
  num get sec => rad * 3600 * 180 / math.pi;

  /// Constructs an HourAngle value corresponding to this angle.
  ///
  /// As both types represent angles in radians, this is a zero-cost conversion.
  HourAngle hourAngle() => HourAngle(rad);

  /// Constructs a Time value where one circle of Angle corresponds to
  /// one day of Time.
  Time time() => Time.fromRad(rad);

  // ---------- Angle math: ----------

  /// Returns the scalar product angle*f
  Angle mul(num f) => Angle(rad * f);

  /// Returns the scalar quotient angle/d
  Angle div(num d) => Angle(rad / d);

  /// Returns Angle a wrapped to 1 circle.
  Angle mod1() => Angle(rad % (2 * math.pi));

  /// Returns the trigonometric sine of this angle.
  num sin() => math.sin(rad);

  /// Returns the trigonometric cosine of this angle.
  num cos() => math.cos(rad);

  /// Returns the trigonometric tangent of this angle.
  num tan() => math.tan(rad);

  /// Returns the absolute value of this angle.
  Angle abs() => Angle(rad.abs());

  // ---------- Simple number math: ----------

  Angle operator +(Angle other) => Angle(rad + other.rad);
  Angle operator -(Angle other) => Angle(rad - other.rad);
  Angle operator *(num scalar) => Angle(rad * scalar);
  Angle operator /(num scalar) => Angle(rad / scalar);

  bool operator <(Angle other) => rad < other.rad;
  bool operator <=(Angle other) => rad <= other.rad;
  bool operator >(Angle other) => rad > other.rad;
  bool operator >=(Angle other) => rad >= other.rad;

  // ---------- Standard overrides: ----------

  @override
  bool operator ==(other) => other is Angle && rad >= other.rad;

  @override
  int get hashCode => rad.hashCode;

  @override
  String toString() => rad.toString();
}

/// An angle corresponding to angular rotation of the Earth.
///
/// The value is stored as radians.
class HourAngle {
  /// The hour angle as an angle in radians.
  ///
  /// This is the underlying representation and involves no scaling.
  final num rad;

  const HourAngle(this.rad);

  /// Constructs an HourAngle value from a value representing
  /// hours of rotation or revolution where there are 24 hours to a revolution.
  const HourAngle.fromHour(num h) : this(h / 12 * math.pi);
  // 12 hours or pi radians in a half-revolution

  /// Constructs an HourAngle value from a value representing
  /// minutes of revolution where there are 60 minutes to an hour and 24 hours
  /// to a revolution.
  const HourAngle.fromMin(num m) : this(m / 60 / 12 * math.pi);
  // 60 sec in an hour, 12 hours or pi radians in a half-revolution

  /// Constructs an HourAngle value from a value representing
  /// seconds of revolution where there are 60 seconds to a minute, 60 minutes
  /// to an hour, and 24 hours to a revolution.
  const HourAngle.fromSec(num s) : this(s / 3600 / 12 * math.pi);
  // 3600 sec in an hour, 12 hours or pi radians in a half-revolution

  /// Constructs a new Angle value from sign, degree, minute, and second
  /// components.
  ///
  /// For argument [neg], pass '-' to negate the result.  Any other argument
  /// value, such as ' ', '+', or simply 0, leaves the result non-negated.
  //
  // This was the NewHourAngle function in the original Go code.
  HourAngle.fromSexa(dynamic neg, int h, int m, num s)
      : this(fromSexa(neg, h, m, s) / 12 * math.pi);

  // ---------- HourAngle "getters" or conversions: ----------

  /// The hour angle as hours of revolution.
  num get hour => rad * 12 / math.pi;

  /// The hour angle as minutes of revolution.
  num get min => rad * 60 * 12 / math.pi;

  /// The hour angle as seconds of revolution.
  num get sec => rad * 3600 * 12 / math.pi;

  /// Returns an Angle value where one revolution or 24 hours of HourAngle
  /// corresponds to one circle of Angle.
  Angle angle() => Angle(rad);

  /// Returns a Time value where one revolution or 24 hours or HourAngle
  /// corresponds to one day of Time.
  Time time() => Time(sec);

  // ---------- Hour angle math: ----------

  /// Returns the scalar product angle*f
  HourAngle mul(num f) => HourAngle(rad * f);

  /// Returns the scalar quotient angle/d
  HourAngle div(num f) => HourAngle(rad / f);

  /// Returns the trigonometric sine of this angle.
  num sin() => math.sin(rad);

  /// Returns the trigonometric cosine of this angle.
  num cos() => math.cos(rad);

  /// Returns the trigonometric tangent of this angle.
  num tan() => math.tan(rad);

  /// Returns the absolute value of this angle.
  HourAngle abs() => HourAngle(rad.abs());

  // ---------- Simple number math: ----------

  HourAngle operator +(Angle other) => HourAngle(rad + other.rad);
  HourAngle operator -(Angle other) => HourAngle(rad - other.rad);
  HourAngle operator *(num scalar) => HourAngle(rad * scalar);
  HourAngle operator /(num scalar) => HourAngle(rad / scalar);

  bool operator <(HourAngle other) => rad < other.rad;
  bool operator <=(HourAngle other) => rad <= other.rad;
  bool operator >(HourAngle other) => rad > other.rad;
  bool operator >=(HourAngle other) => rad >= other.rad;

  // ---------- Standard overrides: ----------

  @override
  bool operator ==(other) => other is HourAngle && rad >= other.rad;

  @override
  int get hashCode => rad.hashCode;

  @override
  String toString() => rad.toString();
}

/// A duration or relative time.
///
/// The value is stored as seconds.
class Time {
  /// The time in seconds.
  ///
  /// This is the underlying representation and involves no scaling.
  final num sec;

  // ---------- Time constructors: ----------

  /// Create a new time given number of seconds
  const Time(num seconds) : sec = seconds;

  /// Constructs a new Time value from sign, hour, minute, and
  /// second components.
  ///
  /// For argument [neg], pass '-' to negate the result.  Any other argument
  /// value, such as ' ', '+', or simply 0, leaves the result non-negated.
  //

  const Time.fromSexa(dynamic neg, int h, int m, num s)
      : this((s + (h * 60 + m) * 60) * (neg == '-' ? -1 : 1));

  /// Constructs a Time value from a value representing days.
  const Time.fromDay(num d)
      : this(d * 3600 * 24); // 3600 sec in an hour, 24 hours in a day

  /// Constructs a Time value from a value representing hours
  /// of time.
  const Time.fromHour(num h) : this(h * 3600); // 3600 sec in an hour

  /// Constructs a Time value from a value representing minutes
  /// of time.
  const Time.fromMin(num m) : this(m * 60); // 60 sec in a min

  /// Constructs a Time value from radians where 2 pi radians
  /// corresponds to one day.
  const Time.fromRad(num rad) : this(rad * 3600 * 12 / math.pi);
  // 3600 sec in an hour, 12 hours or pi radians in a half-day

  // ---------- Time "getters" or conversions: ----------

  /// Time in days.
  num get day => sec / 3600 / 24;

  /// Time in hours.
  num get hour => sec / 3600;

  /// Time in minutes.
  num get min => sec / 60;

  /// Returns time in radians, where 1 day = 2 Pi radians of rotation.
  num get rad => sec / 3600 / 12 * math.pi;

  /// Returns time t as an equivalent angle where 1 day = 2 Pi radians.
  Angle angle() => Angle(rad);

  // HourAngle returns time t as an equivalent hour angle where
  // 1 day = 2 Pi radians or 24 hours of HourAngle.
  HourAngle hourAngle() => HourAngle(rad);

  // ---------- Time math: ----------

  /// Returns the scalar product time*f
  Time mul(num f) => Time(sec * f);

  /// Returns the scalar quotient time/d
  Time div(num d) => Time(sec / d);

  /// Returns a new Time wrapped to one day, the range [0,86400) seconds.
  Time mod1() => Time(sec % (3600 * 24));

  // ---------- Simple number math: ----------

  Time operator +(Time other) => Time(sec + other.sec);
  Time operator -(Time other) => Time(sec - other.sec);

  Time operator *(num scalar) => Time(sec * scalar);
  Time operator /(num scalar) => Time(sec / scalar);

  bool operator <(Time other) => sec < other.sec;
  bool operator <=(Time other) => sec <= other.sec;
  bool operator >(Time other) => sec > other.sec;
  bool operator >=(Time other) => sec >= other.sec;

  /// Returns the time formatted as an ISO8601 string.
  String toIso8601String() {
    final secs = sec.truncate();
    final millisFrac = ((sec - secs) * 1000);
    final millis = millisFrac.truncate();
    final micros = ((millisFrac - millis) * 1000).round();
    return Duration(seconds: secs, milliseconds: millis, microseconds: micros)
        .toString();
  }

  // ---------- Standard overrides: ----------

  @override
  bool operator ==(other) => other is Time && sec == other.sec;

  @override
  int get hashCode => sec.hashCode;

  @override
  String toString() => sec.toString();
}
